### PR TITLE
Feature use select for update to prevent lost updates

### PR DESCRIPTION
**Pull request by pawelkisielewicz**
Resolves https://github.com/golemfactory/concent/issues/718

### CODE REVIEW

File: concent_api/core/subtask_helpers.py [hunk 1]
```diff
@@ -5,6 +5,7 @@
 
 from django.db.models           import Q
 from django.utils               import timezone
+import golem_messages
```
Comment (kbeker) [inline concent_api/core/subtask_helpers.py]: We should not import whole library. Import specific thing what you need to use.

File: concent_api/core/tests/test_unit_message_handlers.py [hunk 2]
```diff
@@ -156,3 +157,24 @@ def test_that_validation_raises_error_if_reason_is_none(self):
...
+
+
+class TestGetSubtaskIdsAsList(ConcentIntegrationTestCase):
+
+    def test_that_correct_subtask_should_be_returned_when_golem_message_with_one_subtask_given(self):
+        force_report_computed_task = self._get_deserialized_force_report_computed_task(
+            self._get_deserialized_report_computed_task(subtask_id='xxyyzz')
+        )
+
+        subtasks_list= get_subtask_ids_as_list(force_report_computed_task)
```
Comment (kbeker) [inline concent_api/core/tests/test_unit_message_handlers.py]: Style. There should be space between `=`

File: concent_api/core/subtask_helpers.py [hunk 3]
```diff
@@ -187,3 +188,15 @@ def are_subtask_results_accepted_messages_signed_by_the_same_requestor(subtask_r
         ) for subtask_results_accepted in subtask_results_accepted_list
     )
     return are_all_signed_by_requestor
+
+
+def get_subtask_ids_as_list(client_message: golem_messages.message) -> List[str]:
```
Comment (kbeker) [inline concent_api/core/subtask_helpers.py]: `extract_subtask_ids_from_client_message` won't be better name for this function?

File: concent_api/api-integration-multi-threads-test.py [hunk 4]
```diff
@@ -0,0 +1,206 @@
...
+from api_testing_common import api_request
+from api_testing_common import assert_condition
+from api_testing_common import create_signed_task_to_compute
+from api_testing_common import run_tests
+from api_testing_common import timestamp_to_isoformat
+from common.helpers import get_current_utc_timestamp
+from common.helpers import sign_message
+
+"""
+Important: to correctly run this tests and create your own tests like this please remember:
```
Comment (rwrzesien) [inline concent_api/api-integration-multi-threads-test.py]: Please do not make comments like this, as they are treated like strings. If you want to make multi-line comment on file level, use `#` several times.

File: concent_api/api-integration-multi-threads-test.py [hunk 5]
```diff
@@ -0,0 +1,206 @@
...
+
+
+def get_unique_subtask_id_for_tests() -> str:
+    now = datetime.now()
+    return str(now.year) + str(now.month) + str(now.day) + str(now.hour) + str(now.minute) + str(now.second) + str(
+        now.microsecond)
+
+
+def call_function_in_treads(
+        func,
```
Comment (rwrzesien) [inline concent_api/api-integration-multi-threads-test.py]: Please fix indentation.

File: concent_api/api-integration-multi-threads-test.py [hunk 6]
```diff
@@ -0,0 +1,206 @@
...
+
+REPORT_COMPUTED_TASK_SIZE = 10
+
+responses = []  # type: list
+
+NUMBER_OF_TESTING_THREADS = 3  # changing this variable remember to correct number of expected responses
+MAXIMUM_WAITING_TIME_FOR_ALL_RESPONSES = 3  # seconds
+
+
+def get_unique_subtask_id_for_tests() -> str:
```
Comment (rwrzesien) [inline concent_api/api-integration-multi-threads-test.py]: There is a function which does that, called `get_task_id_and_subtask_id`. It requires `test_id` which should be passed to each test by runner.

File: concent_api/core/subtask_helpers.py [hunk 7]
```diff
@@ -187,3 +188,15 @@ def are_subtask_results_accepted_messages_signed_by_the_same_requestor(subtask_r
...
+    list_of_subtasks = []
+
+    if isinstance(client_message, golem_messages.message. concents.ForcePayment):
```
Comment (kbeker) [inline concent_api/core/subtask_helpers.py]: Between `message.` and `concents.ForcePayment` should be no break :)

File: concent_api/core/subtask_helpers.py [hunk 8]
```diff
@@ -187,3 +188,15 @@ def are_subtask_results_accepted_messages_signed_by_the_same_requestor(subtask_r
...
+        for subtask_results_accepted in client_message.subtask_results_accepted_list:
+            list_of_subtasks.append(subtask_results_accepted.subtask_id)
+    else:
+        list_of_subtasks = [client_message.subtask_id]
+
```
Comment (kbeker) [inline concent_api/core/subtask_helpers.py]: Here could be assert that `len(list_of_subtasks) > 0`

File: concent_api/api-integration-multi-threads-test.py [hunk 9]
```diff
@@ -0,0 +1,206 @@
...
+
+
+def get_force_get_task_result(report_computed_task: ReportComputedTask) -> ForceGetTaskResult:
+    force_get_task_result = message.concents.ForceGetTaskResult(
+        report_computed_task=report_computed_task,
+    )
+    return force_get_task_result
+
+
+class TestApiMultipleRequestsWithExistingSubtask:
```
Comment (rwrzesien) [inline concent_api/api-integration-multi-threads-test.py]: Is making this a class works ok with the `run_tests` ? I mean, are tests as methods gathered correctly from `globals` ? Besides, I think we should stick to one way of doing this if possible, so if there is special reason for making it a class it should be functions too.

File: concent_api/api_testing_common.py [hunk 10]
```diff
@@ -258,7 +260,7 @@ def parse_arguments():
 
 def get_tests_list(patterns, all_objects):
     def _is_a_test(x):
-        return "case_" in x
+        return "case_" in x or "TestApi" in x
```
Comment (rwrzesien) [inline concent_api/api_testing_common.py]: If it is not necessary, please remove this.

File: concent_api/core/subtask_helpers.py (hunk 1 above)
Comment (rwrzesien) [inline concent_api/core/subtask_helpers.py]: Please import only necessary stuff.

File: concent_api/core/subtask_helpers.py (hunk 3 above)
Comment (rwrzesien) [inline concent_api/core/subtask_helpers.py]: Should be `golem_messages.message.Message`, but after import only `Message`.

File: concent_api/core/subtask_helpers.py (hunk 7 above)
Comment (rwrzesien) [inline concent_api/core/subtask_helpers.py]: Typo, extra space after `golem_messages.message.`.

File: concent_api/api-integration-multi-threads-test.py [hunk 11]
```diff
@@ -0,0 +1,206 @@
...
+
+"""
+Important: to correctly run this tests and create your own tests like this please remember:
+1. TestApi should be in all class names
+2. test_that should be in all test methods in classes names
+"""
+
+os.environ.setdefault("DJANGO_SETTINGS_MODULE", "concent_api.settings")
+
+REPORT_COMPUTED_TASK_SIZE = 10
```
Comment (kbeker) [inline concent_api/api-integration-multi-threads-test.py]: If you are creating this new test, please don't hard code this value. Create file and read size from file and then assign it to report_computed_task

File: concent_api/api-integration-multi-threads-test.py [hunk 12]
```diff
@@ -0,0 +1,206 @@
...
+        t.start()
+
+
+def get_force_report_computed_task(
+    task_to_compute: TaskToCompute,
+    report_computed_task: ReportComputedTask,
+) -> ForceReportComputedTask:
+    force_report_computed_task = ForceReportComputedTask()
+    force_report_computed_task.report_computed_task = report_computed_task
+    force_report_computed_task.report_computed_task.task_to_compute = task_to_compute
```
Comment (kbeker) [inline concent_api/api-integration-multi-threads-test.py]: There should not be task_to_compute added to others message this way. Please try to do this like in real flow:
1) Create TaskToCompute and sign it
2) Create ReportComputedTask with signed TaskToCompute
3) Create ForceReportComputedTask with signed ReportComputedTask from point 2

File: concent_api/api-integration-multi-threads-test.py [hunk 13]
```diff
@@ -0,0 +1,206 @@
...
+
+
+def get_ack_report_computed_task(
+    task_to_compute: TaskToCompute,
+    report_computed_task: ReportComputedTask,
+) -> AckReportComputedTask:
+    ack_report_computed_task = AckReportComputedTask()
+    ack_report_computed_task.report_computed_task = report_computed_task
+    ack_report_computed_task.report_computed_task.task_to_compute = task_to_compute
+
```
Comment (kbeker) [inline concent_api/api-integration-multi-threads-test.py]: Same here as above

File: concent_api/api-integration-multi-threads-test.py [hunk 14]
```diff
@@ -0,0 +1,206 @@
...
+responses = []  # type: list
+
+NUMBER_OF_TESTING_THREADS = 3  # changing this variable remember to correct number of expected responses
+MAXIMUM_WAITING_TIME_FOR_ALL_RESPONSES = 3  # seconds
+
+
+def get_unique_subtask_id_for_tests() -> str:
+    now = datetime.now()
+    return str(now.year) + str(now.month) + str(now.day) + str(now.hour) + str(now.minute) + str(now.second) + str(
+        now.microsecond)
```
Comment (kbeker) [inline concent_api/api-integration-multi-threads-test.py]: This not looks good. We should not create ids like this. If you want to create complicate ids there is always this issue to do https://github.com/golemfactory/concent/issues/705 :)

File: concent_api/api-integration-multi-threads-test.py [hunk 15]
```diff
@@ -0,0 +1,206 @@
...
+2. test_that should be in all test methods in classes names
+"""
+
+os.environ.setdefault("DJANGO_SETTINGS_MODULE", "concent_api.settings")
+
+REPORT_COMPUTED_TASK_SIZE = 10
+
+responses = []  # type: list
+
+NUMBER_OF_TESTING_THREADS = 3  # changing this variable remember to correct number of expected responses
```
Comment (kbeker) [inline concent_api/api-integration-multi-threads-test.py]: So please add assert that number of threads == number of expected responses

**Comment by pawelkisielewicz** [comment]
This will be resolved in other pull request, but thanks for your suggestions, I used it for new version

**Pull request closed without merging**